add parallel-safety functionality to run-api-tests for testing parallelization

Smackteo opened the pull request:
#### 94022d76e6266c7acb456b43779eaf36a753cc6d
<pre>
add parallel-safety functionality to run-api-tests for testing parallelization
<a href="https://bugs.webkit.org/show_bug.cgi?id=302018">https://bugs.webkit.org/show_bug.cgi?id=302018</a>
<a href="https://rdar.apple.com/164101543">rdar://164101543</a>

Reviewed by Jonathan Bedard.

This adds functionality for testing parallelization on a given test.

* Tools/Scripts/run-api-tests:
(parse_args):
* Tools/Scripts/webkitpy/api_tests/manager.py:
(Manager.run):
* Tools/Scripts/webkitpy/api_tests/runner.py:
(run_parallel_safety_single_iteration):
(Runner.run):
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.sharding_groups):

Canonical link: <a href="https://commits.webkit.org/304306@main">https://commits.webkit.org/304306@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04b98f32e9ac714d1b4aff325e74cc27d234d9d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135207 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7599 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142713 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137076 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8235 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7447 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103319 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138153 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5875 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/121187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84178 "Passed tests") | | ⏳ 🛠 vision-apple 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/134555 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3268 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3304 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/39362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145410 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7281 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/39930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111698 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7325 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/6100 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112062 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28431 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5507 "Passed tests") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/117481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7335 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7091 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7311 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7194 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -155,44 +172,69 @@ def run(self, tests, num_workers):
-                    # Dispatch shards from non-system groups first
+                    for test_name in cross_test_tasks:
+                        _log.info('Dispatching infinite cross-test task for {} to cross-test group'.format(test_name))
+                        pool.do(run_cross_test_single_iteration, test_name, infinite=True, group='cross-test')
```
How does this work if the user specifies more tests than we have processes to run them?

Smackteo replied:
This is the last little tweak I had in mind to make.
Essentially wrap everything in a loop and only allow 1/3 of the available processes to be infinite. If more tests are specified it will do those tests in the next loop.

JonWBedard commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
```diff
@@ -296,6 +321,9 @@ def main(cls, name, loglevel, setup, setupargs, setupkwargs, queue, teardown, te
         cls.working = True
+        cls.name = name
+        cls.working = True
+        cls.infinite_task_queue = []
```
This is a big change to `TaskPool`. Which is necessary, I think, to achieve the goal. However, we need to write `TaskPool` unit tests for this, and probably break the `TaskPool` changes into a different PR, since they're substantial.

Smackteo replied:
> This is a big change to `TaskPool`. Which is necessary, I think, to achieve the goal. However, we need to write `TaskPool` unit tests for this, and probably break the `TaskPool` changes into a different PR, since they're substantial.

Understood, completely reasonable and I can do that.

bls1999 commented on Tools/Scripts/run-api-tests:
```diff
@@ -145,6 +145,8 @@ def parse_args(args):
                              help='Force all tests to run in parallel (without grouping shards)'),
         optparse.make_option('--additional-env-var', type='string', action='append', default=[],
                              help='Passes that environment variable to the tests (--additional-env-var=NAME=VALUE)'),
+        optparse.make_option('--cross-test-with', type='string', action='append', default=[],
```
This option name doesn't really make a lot of sense to me, even with the help line. What is this supposed to do and how can the flag name better reflect that?

Smackteo replied:
--cross-test-with means running a single test in repeat while running regularly supplied tests. I'll try to think of a better name

bls1999 replied:
Under what circumstances do you envision this being used?

Smackteo replied:
> Under what circumstances do you envision this being used?

If you run a test against another you can determine if that test can be ran in parallel with that other test. Now run it on repeat against every other test and you've just determined if the test needs to be removed from the allow list for parallel tests. Take it one step further and run it as an EWS step for newly added tests or newly modified tests and we've created a way to determine if a new test is parallel-able.

bls1999 replied:
Discussed offline, `test-parallel-safety` is a better name.

bls1999 commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -45,6 +45,24 @@ def run_shard(name, *tests):
+        _Worker.instance._run_single_test(binary_name, test_suffix)
+        return f"Cross-test iteration completed for {test_name}"
+    except Exception as e:
+        _log.error('Error in cross-test iteration for {}: {}'.format(test_name, e))
```
Nit: you should use an `fstring` here and in a few other places.

Smackteo replied:
Taken care of already locally for the next PR iteration.

bls1999 commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -202,11 +244,14 @@ def run(self, tests, num_workers):
             server_process_logger.setLevel(original_level)
             Runner.instance = None
 
+
```
nit: formatting churn

bls1999 reviewed (commented):
Added a few high-level comments, nothing too major. I share @JonWBedard's sentiment about large changes to TaskPool being particularly disruptive, so please make sure unit tests are passing, and that we test this extensively before merging.

bls1999 commented on Tools/Scripts/run-api-tests:
```diff
@@ -145,6 +145,8 @@ def parse_args(args):
                              help='Force all tests to run in parallel (without grouping shards)'),
         optparse.make_option('--additional-env-var', type='string', action='append', default=[],
                              help='Passes that environment variable to the tests (--additional-env-var=NAME=VALUE)'),
+        optparse.make_option('--test-parallel-safety-with', type='string', action='append', default=[],
```
You may want to drop `-with` from the option name, as the rest of your comments and variables are `test_parallel_safety` sans `with`.

bls1999 commented on Tools/Scripts/webkitpy/api_tests/manager.py:
```diff
@@ -256,7 +256,13 @@ def run(self, args, json_output=None):
+        test_parallel_safety_count = len([test for arg in test_parallel_safety_tasks for test in arg.split()])
+        expected_test_count = len(set(test_names)) + test_parallel_safety_count
+
+        _log.info('Ran {} tests of {} with {} successful'.format(len(runner.results) - disabled, expected_test_count, len(successful)))
```
Nit: this (among other places) can be an f-string

bls1999 commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -45,6 +45,24 @@ def run_shard(name, *tests):
+    # Check if worker is ready - repeat tasks can start before setup completes
+    if _Worker.instance is None:
+        # Worker not ready yet, return early (TaskPool will retry)
+        _log.debug('Worker not ready for {}, skipping iteration'.format(test_name))
```
You write the same string twice here. This could be:
```python3
not_ready = f'Worker not ready for {test_name}, skipping iteration'
_log.debug(not_ready)
return not_ready
```

bls1999 reviewed (approved):
[unofficial] r+ with comments

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -155,53 +173,133 @@ def run(self, tests, num_workers):
+                test_parallel_safety_tasks = supplied_tests
+
+            # Calculate batching for test-parallel-safety tasks to prevent overwhelming workers
+            max_repeat_workers = max(1, workers // 3)  # Reserve 1/3 of workers for repeat tasks
```
Can we put this before the previous `if` and get rid of `test_parallel_safety_tasks`, and just use `test_parallel_safety_batches`?

Smackteo replied:
furthermore the previous if can be combined into this if... making that change.

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -155,53 +173,133 @@ def run(self, tests, num_workers):
+                    _log.info(f'Test-parallel-safety batching: {len(test_parallel_safety_tasks)} tasks split into {len(test_parallel_safety_batches)} batches of max {max_repeat_workers} tasks each')
+
+            # For worker calculation, use current batch size (first batch or all if unbatched)
+            current_test_parallel_safety_count = len(test_parallel_safety_batches[0]) if test_parallel_safety_batches else 0
```
What is the purpose of `current_test_parallel_safety_count`?

Smackteo replied:
it's used later on to calculate the minimum needed workers (if a device can spawn 15 but you only need 3 for a task for example.) This can be combined lower into one variable though.

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -155,53 +173,133 @@ def run(self, tests, num_workers):
-                        pool.do(run_shard, name, *tests)
-
-                    pool.wait()
+            if non_system_shards or test_parallel_safety_batches:
```
Seems like we should be pushing this logic earlier in the function. If we are adding new tests to validate in parallel, but we don't have any non-system shards, isn't this whole function useless?

Smackteo replied:
indeed good catch, fixed.

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -155,7 +176,30 @@ def run(self, tests, num_workers):
+            # Calculate batching for test-parallel-safety tasks to prevent overwhelming workers
+            if supplied_tests_raw:
+                for test_arg in supplied_tests_raw:
+                    supplied_tests.extend(test_arg.split())
```
Why are we splitting the `test_arg`? Those should be single test names.

Smackteo replied:
`self.port.get_option('test_parallel_safety')` gives the test names in `['TestWebKitAPI.DFACombinerTest.Basic TestWebKitAPI.WKWebExtension.SymbolImageIcon']` format if multiple tests are supplied in the cli.
that makes `test_arg 'TestWebKitAPI.DFACombinerTest.Basic TestWebKitAPI.WKWebExtension.SymbolImageIcon'`

so splitting it gives the test names individually

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -166,33 +210,76 @@ def run(self, tests, num_workers):
+                        pool.wait()
+
+                # Handle case with no test-parallel-safety tasks but regular shards exist
+                if not test_parallel_safety_batches and non_system_shards:
```
I still don't understand why we need two different `if` statements here. First, we should use `if/else` if we really do need that. Second, it seems to me that the only real difference between this statement and the above one is the `for i, test_name in enumerate(test_parallel_safety_batch):`, can we not have them share everything and then keep that one `for` loop inside an `if` statement conditionalized on `test_parallel_safety_batch`?

Smackteo replied:
I think my original thought was to leave the regular task pool creation untouched for ease of future modification, but I agree with your points looking at it now. Made this change.

Smackteo committed:
add parallel-safety functionality to run-api-tests for testing parallelization
https://bugs.webkit.org/show_bug.cgi?id=302018
rdar://164101543

Reviewed by Jonathan Bedard.

This adds functionality for testing parallelization on a given test.

* Tools/Scripts/run-api-tests:
(parse_args):
* Tools/Scripts/webkitpy/api_tests/manager.py:
(Manager.run):
* Tools/Scripts/webkitpy/api_tests/runner.py:
(run_parallel_safety_single_iteration):
(Runner.run):
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.sharding_groups):

Canonical link: https://commits.webkit.org/304306@main
